design/ironic-standalone-operator.md

```diff
@@ -0,0 +1,369 @@
+
+Using a provisioning network will require having a provisioning IP per each
+control plane node instead of only one with the non-HA architecture.
+
+Using iPXE in the HA configuration poses one more problem. Our (static) DHCP
+configuration must point each host at its iPXE configuration script. However,
+dnsmasq does not know, which host belongs to which Ironic instance. To tackle
+this limitation, a new [boot configuration API][boot config] has been proposed
+(but not yet implemented) in Ironic. It will allow our DHCP configuration to
+always point at the same Ironic instance for iPXE configuration, and Ironic
+itself will do the required routing.
+
+[boot config]: https://specs.openstack.org/openstack/ironic-specs/specs/approved/boot-config-api.html
+
+##### JSON RPC
+
+Ironic itself is a clustered software. Each instance, as noted above, will
+handle its share of all nodes. When an instance crashes, the remaining
+instances will take over its responsibilities. You can hit the API on any
+instance for any node, and the request will be forwarded to the right instance.
+
+To achieve that, Ironic supports JSON RPC. Metal3 currently does not use it,
+and it still will not be used in the non-HA case. For JSON RPC to be usable,
+each Ironic instance must register its RPC access IP or hostname in the
+database.
+
+When TLS support is enabled, the RPC communication must be secured by
+TLS as well. This may pose a problem since each Ironic instance needs a TLS
+certificate that is valid for its RPC access IP or hostname. This problem has
+been extensively discussed in the [initial HA proposal][issue3], and here is
+the proposed solution, at least for the MVP case:
+
+The Ironic controller will generate a self-signed CA and pass its public and
+private parts into each Ironic container. Each Ironic container will generate
+its private key certificate and sign the certificate with this CA. The CA will
+be trusted **only** for the RPC purpose, removing the possibility of abuse.
+To reduce the number of code paths, this process will happen unconditionally,
+even when TLS for Ironic itself is not enabled.
+
+[kea]: https://www.isc.org/kea/
+[issue3]: https://github.com/metal3-io/ironic-standalone-operator/issues/3
+
+#### Architecture FAQ
+
+Q: Why cannot we split dnsmasq into a separate deployment in the non-HA
+architecture? A: That may require having more than one IP address on the
+provisioning network: for dnsmasq and for httpd/ironic. This is a new
+operational requirement that I'd like to avoid at this stage.
+
+Q: Why does the same dnsmasq limitation affect the HA architecture. A: The HA
+architecture is completely new here, so we can introduce new requirements
+without regression in the operational experience.
+
+Q: Why using *DaemonSets* if *StatefulSets* provide us an easier way to address
+separate Ironic instances? A: While we're relying on host networking, making
+several Ironic instances co-exist on the same Kubernetes node is too complex.
```

Comment:
From Kubernetes HA perspective it is also more reliable to have the ironic pods present on separate control plane nodes. In other words clustering all Ironic pods on e.g. 1  controlplane node wouldn't be HA anyways as that node could be lost.